activerecord/lib/active_record/attributes.rb

```diff
@@ -205,14 +205,21 @@ module ClassMethods
       # tracking is performed. The methods +changed?+ and +changed_in_place?+
       # will be called from ActiveModel::Dirty. See the documentation for those
       # methods in ActiveModel::Type::Value for more details.
-      def attribute(name, cast_type = nil, **options, &block)
+      def attribute(name, cast_type = nil, **options, &_decorator)
         name = name.to_s
         reload_schema_from_cache
 
-        self.attributes_to_define_after_schema_loads =
-          attributes_to_define_after_schema_loads.merge(
-            name => [cast_type || block, options]
-          )
+        prev_cast_type, prev_options, prev_decorator = attributes_to_define_after_schema_loads[name]
+
+        unless cast_type && prev_cast_type
+          cast_type ||= prev_cast_type
+          options = prev_options || options if options.empty?
+          _decorator ||= prev_decorator
+        end
+
+        self.attributes_to_define_after_schema_loads = attributes_to_define_after_schema_loads.merge(
+          name => [cast_type, options, _decorator]
+        )
```

Comment:
Would this be clearer if we had a separate setter method?

```ruby
def attribute(name, cast_type = nil, **options, &decorator)
  name = name.to_s
  reload_schema_from_cache

  override_attribute_defined_after_schema_load name do |prev_cast_type, prev_options, prev_decorator|
    [ cast_type || prev_cast_type,
      options.presence || prev_options,
      decorator || prev_decorator ] unless cast_type && prev_cast_type
  end
end
```

Or potentially the more wasteful option (hey zip!):

```ruby
def attribute(name, cast_type = nil, **options, &decorator)
  name = name.to_s
  reload_schema_from_cache

  override_attribute_defined_after_schema_load name do |previous_definition|
    [ cast_type, options, decorator ].zip(previous_definition).map { |option| option.find(&:present?) }
  end
end
```

Reply:
This intersects a bit with #39487.  After #39487, the `merge` would become:

```ruby
update_heritable_value_of(:attributes_to_define_after_schema_loads, name, [cast_type, options, decorator])
```